File: translated/f-to-c/DRB108-atomic-orig-no.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//Test if atomic can be recognized properly. No data races.

#include <omp.h>
#include <stdio.h>

int main()
{
    int a;
    a = 0;

    #pragma omp parallel
    {
        #pragma omp atomic
        a = a + 1;
    }

    printf("a = %d\n", a);

    return 0;
} 

File: translated/f-to-go-automodel-bad/DRB108-atomic-orig-no.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//Test if atomic can be recognized properly. No data races.

package main

import (
	"fmt"
	"runtime"
	"sync"
	"sync/atomic"
)

func main() {
	var a int32
	a = 0

	//$omp parallel
	var wg sync.WaitGroup
	numCPU := runtime.NumCPU()
	for i := 0; i < numCPU; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			//$omp atomic
			atomic.AddInt32(&a, 1)
			//$omp end atomic
		}()
	}
	wg.Wait()
	//$omp end parallel

	fmt.Printf("a =%8d\n", a)
}
